cms/djangoapps/contentstore/models.py

```diff
@@ -197,70 +261,90 @@ def update_or_create(
+    """
+    upstream_container = models.ForeignKey(
+        Container,
+        on_delete=models.SET_NULL,
+        related_name="links",
+        null=True,
+        blank=True,
+    )
+    upstream_container_key = ContainerKeyField(
+        max_length=255,
+        help_text=_(
+            "Upstream block key (e.g. lct:...), this value cannot be null "
+            "and is useful to track upstream library blocks that do not exist yet "
+            "or were deleted."
         )
+    )
+
+    class Meta:
+        verbose_name = _("Container Link")
+        verbose_name_plural = _("Container Links")
+
+    def __str__(self):
+        return f"ContainerLink<{self.upstream_container_key}->{self.downstream_usage_key}>"
 
     @classmethod
-    def summarize_by_downstream_context(cls, downstream_context_key: CourseKey) -> QuerySet:
+    def update_or_create(
+        cls,
+        upstream_container: Container | None,
+        /,
+        upstream_container_key: LibraryContainerLocator,
+        upstream_context_key: str,
+        downstream_usage_key: UsageKey,
+        downstream_context_key: CourseKey,
+        version_synced: int,
+        version_declined: int | None = None,
+        created: datetime | None = None,
+    ) -> "ContainerLink":
         """
-        Returns a summary of links by upstream context for given downstream_context_key.
-        Example:
-        [
-            {
-                "upstream_context_title": "CS problems 3",
-                "upstream_context_key": "lib:OpenedX:CSPROB3",
-                "ready_to_sync_count": 11,
-                "total_count": 14
-            },
-            {
-                "upstream_context_title": "CS problems 2",
-                "upstream_context_key": "lib:OpenedX:CSPROB2",
-                "ready_to_sync_count": 15,
-                "total_count": 24
-            },
-        ]
+        Update or create entity link. This will only update `updated` field if something has changed.
         """
-        result = cls.filter_links(downstream_context_key=downstream_context_key).values(
-            "upstream_context_key",
-            upstream_context_title=F("upstream_block__learning_package__title"),
-        ).annotate(
-            ready_to_sync_count=Count("id", Q(ready_to_sync=True)),
-            total_count=Count('id')
-        )
-        return result
+        if not created:
+            created = datetime.now(tz=timezone.utc)
+        new_values = {
+            'upstream_container_key': upstream_container_key,
+            'upstream_context_key': upstream_context_key,
+            'downstream_usage_key': downstream_usage_key,
+            'downstream_context_key': downstream_context_key,
+            'version_synced': version_synced,
+            'version_declined': version_declined,
+        }
+        if upstream_container:
+            new_values.update(
+                {
+                    'upstream_container': upstream_container,
+                }
+            )
```

Comment:
Just for my understanding, this is equivalent to:

```python
new_values['upstream_container'] = upstream_container
```

?

Reply:
Yeah, as far as I can see. I suspect there were multiple fields being added there before some refactor.

Reply:
Actually this code is also unchanged from what's already in `master`

Reply:
@bradenmacdonald Yes, I think while writing this, I had multiple fields but were removed while refactoring. I'll update it in a follow up PR.